cmd/swarm/upload.go

```diff
@@ -30,49 +30,101 @@ import (
 	"path"
 	"path/filepath"
 	"strings"
+	"sync"
 
 	"github.com/ethereum/go-ethereum/cmd/utils"
 	"github.com/ethereum/go-ethereum/log"
 	"gopkg.in/urfave/cli.v1"
 )
 
+func getStdin(b []byte) (int, error) {
+	s := os.Stdin
+	return s.Read(b)
+}
+
 func upload(ctx *cli.Context) {
+
 	args := ctx.Args()
 	var (
 		bzzapi       = strings.TrimRight(ctx.GlobalString(SwarmApiFlag.Name), "/")
 		recursive    = ctx.GlobalBool(SwarmRecursiveUploadFlag.Name)
 		wantManifest = ctx.GlobalBoolT(SwarmWantManifestFlag.Name)
 		defaultPath  = ctx.GlobalString(SwarmUploadDefaultPath.Name)
+		fromStdin    = ctx.GlobalBool(SwarmUpFromStdinFlag.Name)
 	)
+
+	var client = &client{api: bzzapi}
+
+	var entry manifestEntry
+
+	data := make([]byte, 1024*1000)
+	datalength := int(0)
+
 	if len(args) != 1 {
-		utils.Fatalf("Need filename as the first and only argument")
+		if fromStdin {
+			var err error
+			datalength, err = getStdin(data)
```

Comment:
Rather than reading the data into memory, how about writing it to a temporary file and then uploading the file once stdin has been fully consumed? Perhaps use an [`io.LimitReader`](https://golang.org/pkg/io/#LimitedReader) if you still want to limit the size.